server/mqtt.go

```diff
@@ -3140,44 +3298,231 @@ func mqttPubTrace(pp *mqttPublish) string {
-	msgToSend := bb.Bytes()
+func (s *Server) mqttProcessPub(c *client, pp *mqttPublish, trace bool) error {
+	qos := mqttGetQoS(pp.flags)
 
-	var err error
-	// Unless we have a publish permission error, if the message is QoS1, then we
-	// need to store the message (and deliver it to JS durable consumers).
-	if _, permIssue := c.processInboundClientMsg(msgToSend); !permIssue && mqttGetQoS(pp.flags) > 0 {
-		// We need to call flushClients now since this we may have called c.addToPCD
-		// with destination clients (possibly a route). Without calling flushClients
-		// the following call may then be stuck waiting for a reply that may never
-		// come because the destination is not flushed (due to c.out.fsp > 0,
-		// see addToPCD and writeLoop for details).
-		c.flushClients(0)
-		// Store this QoS1 message.
-		_, err = c.mqtt.sess.jsa.storeMsg(mqttStreamSubjectPrefix+string(c.pa.subject), c.pa.hdr, msgToSend)
-	}
-	c.pa.subject, c.pa.mapped, c.pa.hdr, c.pa.size, c.pa.szb, c.pa.reply = nil, nil, -1, 0, nil, nil
+	switch qos {
+	case 0, 1:
+		err := s.mqttInitiateMsgDelivery(c, pp)
+		if qos == 1 {
+			// Send a PUBACK back to the client.
+			c.mqttEnqueuePubResponse(mqttPacketPubAck, pp.pi, trace)
+		}
+		return err
+
+	case 2:
+		err := s.mqttStoreQOS2Msg(c, pp)
+		// Send a PUBREC back to the client.
+		c.mqttEnqueuePubResponse(mqttPacketPubRec, pp.pi, trace)
+		return err
+
+	default:
+		return fmt.Errorf("unreachable: invalid QoS in mqttProcessPub: %v", qos)
+	}
+}
+
+func (s *Server) mqttInitiateMsgDelivery(c *client, pp *mqttPublish) error {
+	natsMsg, headerLen := mqttNewDeliverableMessage(pp, false)
+
+	// Broadcast the message just as we do for QoS0, but this message is
+	// recognized as QoS1 (stored in the NATS header), so it will be ignored
+	// by the QoS1+ subscriptions. Those will expect the message to be
+	// delivered via JetStream.
+	permIssue := s.mqttDeliverNATS(c, pp, natsMsg, headerLen)
+	if permIssue {
+		return nil
+	}
+
+	// <>/<> what is addToPCD?
+	// We need to call flushClients now since this we may have called c.addToPCD
+	// with destination clients (possibly a route). Without calling flushClients
+	// the following call may then be stuck waiting for a reply that may never
+	// come because the destination is not flushed (due to c.out.fsp > 0,
+	// see addToPCD and writeLoop for details).
+	c.flushClients(0)
+
+	// Store the message in JetStream as "$MQTT.msgs.<delivery-subject>"
+	_, err := c.mqtt.sess.jsa.storeMsg(mqttStreamSubjectPrefix+string(pp.subject), headerLen, natsMsg)
 	return err
 }
 
+func (s *Server) mqttStoreQOS2Msg(c *client, pp *mqttPublish) error {
+	// `true` means encode the MQTT PUBLISH packet in the NATS message header.
+	natsMsg, headerLen := mqttNewDeliverableMessage(pp, true)
+
+	// Do not broadcast the message until it has been deduplicated and
+	// released by the sender. Instead store this QoS2 message as
+	// "$MQTT.qos2.<domain>.<account-name>.<client-id>.<PI>". If the message
```

Comment:
Domain should be removed, and all this runs under a given account already, so really client-id and PI.